compiler/src/module.rs

```diff
@@ -0,0 +1,295 @@
+use itertools::Itertools;
+use lsp_types::Url;
+use salsa::query_group;
+use std::{
+    collections::hash_map::DefaultHasher,
+    fmt::{self, Display, Formatter},
+    fs,
+    hash::{Hash, Hasher},
+    path::PathBuf,
+    sync::Arc,
+};
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
+pub struct Module {
+    pub package: Package,
+    pub path: Vec<String>,
+    pub kind: ModuleKind,
+}
+#[derive(Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
+pub enum ModuleKind {
+    Code,
+    Asset,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
+pub enum Package {
+    /// A package written by the user.
+    User(PathBuf),
+
+    /// A package managed by the Candy tooling. This is in some special cache
+    /// directory where `use`d packages are downloaded to.
+    //
+    // TODO: Change this to just storing the package name or something like
+    // that so that the root of the cached packages folder isn't stored
+    // everywhere.
+    External(PathBuf),
+
+    /// An anonymous package. This is created for single untitled files that are
+    /// not yet persisted to disk (such as when opening a new VSCode tab and
+    /// typing some code).
+    ///
+    /// For now, this option is also used for files picked from the file system
+    /// that are not part of the current working directory.
+    //
+    // TODO: Maybe add some sort of package indicator file after all so that we
+    // can allow arbitrary opened files from the file system to access parent
+    // and sibling modules if they're actually part of a larger package.
+    Anonymous { root_content: String },
+}
+
+impl Module {
+    pub fn from_anonymous_content(content: String, kind: ModuleKind) -> Self {
+        Module {
+            package: Package::Anonymous {
+                root_content: content,
+            },
+            path: vec![],
+            kind,
+        }
+    }
+
+    pub fn from_package_root_and_url(package_root: PathBuf, url: Url, kind: ModuleKind) -> Self {
+        match url.scheme() {
+            "file" => {
+                Module::from_package_root_and_file(package_root, url.to_file_path().unwrap(), kind)
+            }
+            "untitled" => Module::from_anonymous_content(
+                url.to_string()["untitled:".len()..].to_owned(),
```

Comment:
```suggestion
                url.to_string().strip_prefix("untitled:").unwrap().to_owned(),
```

But I think this should actually be the file's content and not its URI